src/System.Console/src/System/Console.cs

```diff
@@ -45,6 +50,70 @@ public static TextReader In
             }
         }
 
+        public static Encoding InputEncoding
+        {
+            get
+            {
+                return Volatile.Read(ref _inputEncoding) ?? EnsureInitialized(ref _inputEncoding, () =>
+                {
+                    return ConsolePal.InputEncoding;
+                });
+            }
+            set
+            {
+                CheckNonNull(value, "value");
+                lock (InternalSyncObject)
+                {
+                    // Set the terminal console encoding.
+                    ConsolePal.SetConsoleInputEncoding(value);
+
+                    Volatile.Write(ref _inputEncoding, value);
+
+                    // We need to reinitialize Console.In in the next call to _in
+                    // This will discard the current StreamReader, potentially 
+                    // losing buffered data
+                    Volatile.Write(ref _in, null);
+                }
+            }
+        }
+
+        public static Encoding OutputEncoding
+        {
+            get
+            {
+                return Volatile.Read(ref _outputEncoding) ?? EnsureInitialized(ref _outputEncoding, () =>
+                {
+                    return ConsolePal.OutputEncoding;
+                });
+            }
+            set
+            {
+                CheckNonNull(value, "value");
+
+                lock (InternalSyncObject)
+                {
+                    // Set the terminal console encoding.
+                    ConsolePal.SetConsoleOutputEncoding(value);
+
+                    // Before changing the code page we need to flush the data 
+                    // if Out hasn't been redirected. Also, have the next call to  
+                    // _out reinitialize the console code page.
+                    if (Volatile.Read(ref _out) != null && !_isOutTextWriterRedirected)
```

Comment:
Not sure. I tried tracing back to when this was added but could not find much context. The thing that I found further absurd was that even though SetIn() and SetOut() ultimately get set to console I/p and o/p, it will still assume it to be redirected. Note (_isXXXTextWriterRedirected is set to true the moment we use Setxxx). But changing this behavior would be an AppCompat issue so I let it be.